packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -127,21 +135,123 @@ export function publish(options: InitOptions): void {
+                  const node = mutation.removedNodes[i];
+                  if (node.nodeType === Node.ELEMENT_NODE) {
+                    const childElement = node as Element;
+                    const surfaceDataList = observedRoots.get(childElement);
+
+                    // Check if this was a child we were observing
+                    if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+                      console.log('Child removed', childElement);
+
+                      // Stop observing the removed child
+                      observer.unobserve(childElement);
+                      observedRoots.delete(childElement, surfaceData);
+                      surfaceDataRoots.delete(surfaceData, childElement);
+
+                      // Start observing the parent element again
+                      observer.observe(element);
+                      observedRoots.set(element, surfaceData);
+                      surfaceDataRoots.set(surfaceData, element);
+                    }
+                  }
+                }
+              }
+            }
+
+            // Then handle added nodes
+            let hasViableChild = false;
+            if (mutation.addedNodes.length > 0) {
+              // Check if any of the added nodes are elements we can observe
+              for (let i = 0; i < mutation.addedNodes.length; i++) {
+                const node = mutation.addedNodes[i];
+                if (node.nodeType === Node.ELEMENT_NODE) {
+                  const childElement = node as Element;
+                  console.log('Child found', childElement);
+
+                  if (!ALSurfaceUtils.isSurfaceWrapper(childElement)) {
+                    // Check if we're currently observing the parent element
+                    const surfaceDataList = observedRoots.get(element);
+                    if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+                      // Found a viable child, switch observation to it
+                      observer.unobserve(element);
+                      observer.observe(childElement);
+
+                      // Update tracking maps
+                      observedRoots.delete(element, surfaceData);
+                      observedRoots.set(childElement, surfaceData);
+                      surfaceDataRoots.delete(surfaceData, element);
+                      surfaceDataRoots.set(surfaceData, childElement);
+                      hasViableChild = true;
+                      break;
+                    }
+                  }
+                }
+              }
+            }
+            // If we are tracking removal,  we need to keep the observer around,  which could be potentially expensive
+            if (hasViableChild && !enableDynamicChildTrackingForRemoval) {
+              // Disconnect the mutation observer as we no longer need it
+              mutationObserver.disconnect();
+              mutationObservers.delete(element);
+            }
```

Comment:
If we go with a logic of have one observer and just add/remove nodes to it, do we still need to have separate flags for removal?
I think actually for correctness, we should always removal and do cleanup. 
I wonder in the cases that we do find a root element, what happens if that root element is removed. Is that a leak for us?

Reply:
The concern was leaving resources lying around with just accounting for the delayed `added` case,  but given your other feedback I will take a look at using one mutation observer.

If we need to track removal too we have to keep monitoring the element's childlist and can't clean it up like this block is doing.

Reply:
Once something is added or removed, you should cleanup the observation. Also, when something is unmounted, you should do that. 
You can think of handling these well, and cleanup separately.
Also, I'm not sure if creating too many observers itself is cheaper or any better.